processor/schemaprocessor/internal/migrate/conditional.go

```diff
@@ -0,0 +1,92 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package migrate // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/schemaprocessor/internal/migrate"
+
+import "go.opentelemetry.io/collector/pdata/pcommon"
+
+// ValueMatch defines the expected match type
+// that is used on creation of `ConditionalAttributeSet`
+type ValueMatch interface {
+	~string
+}
+
+type ConditionalAttributeSet struct {
+	on    *map[string]struct{}
+	attrs *AttributeChangeSet
+}
+
+type ConditionalAttributeSetSlice []*ConditionalAttributeSet
+
+func NewConditionalAttributeSet[Key AttributeKey, Value AttributeKey, Match ValueMatch](mappings map[Key]Value, matches ...Match) *ConditionalAttributeSet {
```

Comment:
Same here, does this need to be generic? It seems only string types are ever needed.

Reply:
This one needs to be generic since the schema package defines the [types](go.opentelemetry.io/otel/schema/v1.0/types) package that forces those values to be new types instead of alias for strings.

The generic receiver here means we can take the `can convert to string` values and store them internally as strings with one cast in the constructer. 
Save me having to write optional handlers for each type defined by the `types` package.

Reply:
I see.

Instead of a string can we use `types.AttributeName`, e.g. `map[types.AttributeName]types.AttributeName`? Shouldn't this work without type casts?

I saw your [other PR ](https://github.com/open-telemetry/opentelemetry-go/pull/4082). I think I would prefer to keep the strong types in the `types` package. If it means we need to use generics here then so be it.

Reply:
It looks like I can use `ast.AttributeMap` since that is used every where, but with the conditionals, looks like I would need the generic type for `SpanName`, `EventName`, `MetricName`.